libraries/WebServer/src/Middlewares.h

```diff
@@ -0,0 +1,208 @@
+#ifndef MIDDLEWARES_H
+#define MIDDLEWARES_H
+
+#include <WebServer.h>
+#include <Stream.h>
+
+#include <assert.h>
+
+// curl-like logging middleware
+class LoggingMiddleware : public Middleware {
+private:
+  Stream *_out;
+
+public:
+  explicit LoggingMiddleware(Stream &out) : _out(&out) {}
+
+  bool run(WebServer &server, Middleware::Callback next) override {
+    _out->print(F("* Connection from "));
+    _out->print(server.client().remoteIP().toString());
+    _out->print(F(":"));
+    _out->println(server.client().remotePort());
+
+    _out->print(F("< "));
+    HTTPMethod method = server.method();
+    if (method == HTTP_ANY) {
+      _out->print(F("HTTP_ANY"));
+    } else {
+      _out->print(http_method_str((http_method)method));
+    }
+    _out->print(F(" "));
+    _out->print(server.uri());
+    _out->print(F(" "));
+    _out->println(server.version());
+
+    int n = server.headers();
+    for (int i = 0; i < n; i++) {
+      String v = server.header(i);
+      if (!v.isEmpty()) {
+        // because these 2 are always there, eventually empty: "Authorization", "If-None-Match"
+        _out->print(F("< "));
+        _out->print(server.headerName(i));
+        _out->print(F(": "));
+        _out->println(server.header(i));
+      }
+    }
+
+    _out->println(F("<"));
+
+    bool ret = next();
+
+    if (ret) {
+      _out->println(F("* Processed!"));
+
+      _out->print(F("> "));
+      _out->print(F("HTTP/1."));
+      _out->print(server.version());
+      _out->print(F(" "));
+      _out->print(server.responseCode());
+      _out->print(F(" "));
+      _out->println(WebServer::responseCodeToString(server.responseCode()));
+
+      n = server.responseHeaders();
+      for (int i = 0; i < n; i++) {
+        _out->print(F("> "));
+        _out->print(server.responseHeaderName(i));
+        _out->print(F(": "));
+        _out->println(server.responseHeader(i));
+      }
+
+      _out->println(F(">"));
+
+    } else {
+      _out->println(F("* Not processed!"));
+    }
+
+    return ret;
+  }
+};
+
+class AuthenticationMiddleware : public Middleware {
+private:
+  // authenticate state
```

Comment:
replace with enum class instead of comments?
from
```
// authenticate state
// 0: not authenticated
// 1: callback
// 2: username/password
// 3: sha1
```
to
```
enum class AuthState : std::int8_t {
             NOT_AUTHENTICATED, // 0: not authenticated
             CALLBACK, // 1: callback
             USERNAME_PASSWD, // 2: username/password
             SHA1// 3: sha1     
};
```       
?